couchdb/server/compiler.py

```diff
@@ -297,3 +319,69 @@ def require(path, module=None):
                 _visited_ids.pop()
 
     return require
+
+
+def compile_func(funsrc, ddoc=None, context=None, encoding='utf-8', **options):
+    """Compile source code and extract function object from it.
+
+    :param funsrc: Python source code.
+    :type funsrc: unicode
+
+    :param ddoc: Optional argument which must represent design document.
+    :type ddoc: dict
+
+    :param context: Custom context objects which function could operate with.
+    :type context: dict
+
+    :param options: Compiler config options.
+
+    :param encoding: Encoding of source code.
+    :type encoding: str
+
+    :return: Function object.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If source code compilation failed or it doesn't contains function
+          definition.
+
+    .. note::
+        ``funsrc`` should contains only one function definition and import
+        statements (optional) or :exc:`~couchdb.server.exceptions.Error`
+        will be raised.
+
+    """
+    if not context:
+        context = DEFAULT_CONTEXT.copy()
+    else:
+        context, _ = DEFAULT_CONTEXT.copy(), context
+        context.update(_)
+    if ddoc is not None:
+        context['require'] = require(ddoc, context, **options)
+
+    globals_ = {}
+    try:
+        bytecode = compile_to_bytecode(funsrc, encoding=encoding)
+        exec(bytecode, context, globals_)
+    except Exception as err:
+        log.exception('Failed to compile source code:\n%s', funsrc)
+        raise Error('compilation_error', str(err))
+
+    msg = None
+    func = None
+    for item in globals_.values():
+        if isinstance(item, FunctionType):
+            if func is None:
+                func = item
+            else:
+                msg = 'Multiple functions are defined. Only one is allowed.'
+        elif not isinstance(item, ModuleType):
+            msg = 'Only functions could be defined at top level namespace'
```

Comment:
Because there is no any sane reason for using anything else except modules and functions in globals. Otherwise users will try to use classes and hit awkward issues because all the design functions should be pure.